staging/src/k8s.io/apimachinery/pkg/watch/mux.go

```diff
@@ -210,19 +210,38 @@ func (m *Broadcaster) closeAll() {
 }
 
 // Action distributes the given event among all watchers.
-func (m *Broadcaster) Action(action EventType, obj runtime.Object) {
+func (m *Broadcaster) Action(action EventType, obj runtime.Object) error {
+	m.incomingBlock.Lock()
+	defer m.incomingBlock.Unlock()
+	select {
+	case <-m.stopped:
+		return fmt.Errorf("broadcaster already stopped")
+	default:
+	}
+
 	m.incoming <- Event{action, obj}
+	return nil
```

Comment:
can this deadlock?
we are holding the lock and waiting for the incoming channel to be drained ...

Reply:
This seems fine to me, because taking items from `m.incoming` doesn't require a lock (a lock is rather taken to ensure it will not get stopped in the meantime).

So `m.loop` is taking items from m.incoming and is lock-free - thsi looks fine to me.

Reply:
it seems we can't , `go m.loop()` should be consuming the events always and the channel will not be closed because we have the lock

Reply:
Let's try to avoid new functions by now, that change will not be backportable and it may cause issues with downstream , think in Openshift per example, we always move slower with the tool chain